src/test.cpp

```diff
@@ -0,0 +1,297 @@
+        for (size_t impl = 0; impl < sketches.size(); ++impl) {
+            CHECK(sketches[impl].GetBits() == bits);
+            CHECK(sketches[impl].GetCapacity() == capacity);
+            CHECK(sketches[impl].GetSerializedSize() == sketches[0].GetSerializedSize());
+        }
+        if (sketches.empty()) return;
+        // Determine the number of elements, and create a vector to store them in.
+        size_t element_count = std::max<int64_t>(0, std::max<int64_t>(0, capacity + size_offset_dist(rnd)));
+        elements.resize(element_count);
+        // Add the elements to all sketches
+        for (size_t j = 0; j < element_count; ++j) {
+            uint64_t elem = element_dist(rnd);
+            CHECK(elem != 0);
+            elements[j] = elem;
+            for (auto& sketch : sketches) sketch.Add(elem);
+        }
+        // Remove pairs of duplicates in elements, as they cancel out.
+        std::sort(elements.begin(), elements.end());
+        size_t real_element_count = element_count;
+        for (size_t pos = 0; pos + 1 < elements.size(); ++pos) {
+            if (elements[pos] == elements[pos + 1]) {
+                real_element_count -= 2;
+                // Set both elements to 0; afterwards we will move these to the end.
+                elements[pos] = 0;
+                elements[pos + 1] = 0;
+                ++pos;
+            }
+        }
+        if (real_element_count < element_count) {
+            // Move all introduced zeroes (masking duplicates) to the end.
+            std::sort(elements.begin(), elements.end(), [](uint64_t a, uint64_t b) { return a != b && (b == 0 || (a != 0 && a < b)); });
+            CHECK(elements[real_element_count] == 0);
+            elements.resize(real_element_count);
+        }
+        // Create and compare serializations
+        auto serialized_0 = sketches[0].Serialize();
+        for (size_t impl = 1; impl < sketches.size(); ++impl) {
+            auto serialized_other = sketches[impl].Serialize();
+            CHECK(serialized_other == serialized_0);
+        }
+        // Deserialize and reserialize them
+        for (size_t impl = 0; impl < sketches.size(); ++impl) {
+            sketches[impl].Deserialize(serialized_0);
+            auto reserialized = sketches[impl].Serialize();
+            CHECK(reserialized == serialized_0);
+        }
+        // Decode and compare results
+        decode_0.resize(capacity);
+        bool decodable_0 = sketches[0].Decode(decode_0);
+        std::sort(decode_0.begin(), decode_0.end());
+        for (size_t impl = 1; impl < sketches.size(); ++impl) {
+            decode_other.resize(capacity);
+            bool decodable_other = sketches[impl].Decode(decode_other);
+            CHECK(decodable_other == decodable_0);
+            std::sort(decode_other.begin(), decode_other.end());
+            CHECK(decode_other == decode_0);
+        }
+        // If the result is decodable, it should also be decodable with capacity
```

Comment:
"with capacity" is confusing here, makes it sound like you're decoding the sketch at its original capacity.  Maybe "If the result is decodable, it should also be decodable with the actual size even if that is less than capacity but not decodable with one less" or similar?

Reply:
Done (and in a few other places where limit and capacity were being confused).